### PR TITLE
8295125: os::signal should be os specific

### PR DESCRIPTION
**Pull request by dholmes-ora**
This was intended as a significant cleanup RFE but it turned out things were a lot messier than initially envisaged, so the resultant cleanup is quite limited.

The OS API/namespace was simplified by removing three signal related functions that are not needed in shared code:
 - signal
 - user_handler
 - raise
The first two were reworked and placed into PosixSignals and os::win32.
There was no need for a specific XX::raise API as os::raise simply called ::raise.

On the Posix side we had the basic signal handler installation code in two places:

1. Inline in set_signal_handler and used by nearly all the VM signals
2. In os::signal which was used by three different "clients"
  - VM error handler for updating crash handler
  - VM for installing SIG_BREAK handler
  - JDK via JVM_RegisterSignal

This was refactored so that we have:

1. PosixSignals::install_sigaction_signal_handler
  - used by the VM only
    - set_signal_handler
    - VM error handler for updating crash handler
    - VM for installing SIG_BREAK handler

2. PosixSignals::install_generic_signal_handler
  - used by JDK via JVM_RegisterSignal

The sigaction handlers were all consistently defined as per the posix definition:
    void func(int signo, siginfo_t *info, void *context);
  

On the Windows side we just fixed the incorrect definition of UserHandler and made some consistency changes to use the right signal handler type where possible.

Thanks.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8295125](https://bugs.openjdk.org/browse/JDK-8295125): os::signal should be os specific

### Reviewers
 * [Johan Sjölen](https://openjdk.org/census#jsjolen) (@jdksjolen - Committer)
 * [Kim Barrett](https://openjdk.org/census#kbarrett) (@kimbarrett - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10780/head:pull/10780` \
`$ git checkout pull/10780`

Update a local copy of the PR: \
`$ git checkout pull/10780` \
`$ git pull https://git.openjdk.org/jdk pull/10780/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10780`

View PR using the GUI difftool: \
`$ git pr show -t 10780`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10780.diff">https://git.openjdk.org/jdk/pull/10780.diff</a>

</details>

### CODE REVIEW

**Review by jdksjolen — APPROVED** [review]
Thank you for this change, it looks like a good clean up to me.

File: src/hotspot/os/posix/signals_posix.cpp [hunk 1]
```diff
@@ -842,29 +837,68 @@ static bool check_signal_handler(int sig) {
...
 
   return get_signal_handler(&oldSigAct);
 }
 
-void os::signal_raise(int signal_number) {
-  ::raise(signal_number);
+// Installs the given sigaction handler for the given signal.
+// - sigAct: the new struct sigaction to be filled in and used
+//           for this signal. The caller must provide this as it
+//           may need to be stored/accessed by that caller.
```
Comment (kimbarrett) [inline src/hotspot/os/posix/signals_posix.cpp:877]: Maybe I missed something, but I didn't find any places where the caller needs sigAct.

**Review by kimbarrett — CHANGES_REQUESTED** [review]

File: src/hotspot/os/windows/os_windows.hpp [hunk 2]
```diff
@@ -120,6 +122,10 @@ class os::win32 {
     _thread_ptr_offset = offset;
   }
   static inline int get_thread_ptr_offset() { return _thread_ptr_offset; }
+
+  // signal support
+  static void* install_signal_handler(int sig, signal_handler_t handler);
+  static void* user_handler();
```
[the remark this answers is not in the document]
Comment (dholmes-ora) [reply src/hotspot/os/windows/os_windows.hpp:128]: If `user_handler` returned the actual type you would have to cast it to `void*` in the later comparison, just moving the cast in `JVM_RegisterSignal` but saving the cast in `os::win32::user_handler()`. This seemed a small gain compared to maintaining consistency with the Posix version - but I can change it if you want.

File: src/hotspot/os/windows/jvm_windows.cpp [hunk 3]
```diff
@@ -43,11 +44,9 @@ JVM_END
  */
 
 JVM_ENTRY_NO_ENV(void*, JVM_RegisterSignal(jint sig, void* handler))
-  // Copied from classic vm
-  // signals_md.c       1.4 98/08/23
-  void* newHandler = handler == (void *)2
-                   ? os::user_handler()
-                   : handler;
+  signal_handler_t newHandler = handler == (void *)2 ?
+                                CAST_TO_FN_PTR(signal_handler_t, os::win32::user_handler()) :
```
[the remark this answers is not in the document]
Comment (dholmes-ora) [reply src/hotspot/os/windows/jvm_windows.cpp:48]: See previous response.

File: src/hotspot/os/posix/signals_posix.cpp (hunk 1 above)
Comment (dholmes-ora) [reply src/hotspot/os/posix/signals_posix.cpp:877]: It is in `set_signal_handlers`:
```
struct sigaction sigAct;
  int ret = PosixSignals::install_sigaction_signal_handler(&sigAct, &oldAct,
                                                           sig, javaSignalHandler);
  assert(ret == 0, "check");
  // Save handler setup for possible later checking
  vm_handlers.set(sig, &sigAct);
```
and is the reason I pass it in.

**Comment by dholmes-ora** [comment]
Thanks for looking at this @jdksjolen  and @kimbarrett !

File: src/hotspot/os/posix/signals_posix.cpp (hunk 1 above)
Comment (kimbarrett) [reply src/hotspot/os/posix/signals_posix.cpp:877]: Yes, I missed that.

**Review by kimbarrett — APPROVED** [review]
I'm loath to say "looks good", but looks okay.

Fixing JDK-8295702 will likely involve some refactoring of what's here. But
that's a problem for the next person who pokes at this tar baby.

**Comment by dholmes-ora** [comment]
Thanks @kimbarrett !

/integrate

**Pull request closed without merging**